### ISSUE TITLE
Stageout Sometimes Treated as Failure Because of Problem with Remote File Size Check

### ISSUE THREAD

**Issue by klannon** [NDCMS/lobster#626]
I've noticed on the dedicated T3 workers, sometimes stageout is marked as a failure because it tries to use a local operation (i.e. `stat`) to measure the size of the remote file, but the size is returned as "0".

I'm having trouble believing that the size is really zero because I see a consistent pattern of attempt to do a local file copy, measure file size zero, attempt to do a remote copy, and fail because file already exists.  I modified stageout with `gfal-copy` to use `-f` so that it will overwrite files that exist, and it still failed because `check_output` uses all the methods defined in the config to access the output, so if `file:` is before `gsiftp:` then even if it's `gsiftp:` that's used to copy, it will still use `stat` to check the size if it can and fail with size = 0.  The problem isn't consistent in that some jobs fail and some succeed.  I suspect it's an issue with FUSE on some of the hosts.

The solution I'm contemplating is to treat mismatched file sizes as a failure of the file size measurement rather than as a failure of the transfer.  So if the local and remote files don't match, Lobster would try to measure the file size with the next available method.  File transfers would only be counted as failures if all methods of measuring file size failed.

There is one other potential problem here.  `check_output` is coded so that if none of the implemented file size check methods succeed, then it returns `True`.  The comments suggest this is a compromise because there are transfer methods (e.g. `gsiftp:`) which don't have an implemented file measurement technique, so it's possible to construct a config where outputs could be transferred but not have their size checked.  To get around this, I am think I'll implement file size checks using `gfal-stat` and `hadoop dfs -stat "Size: %b"` to close the loopholes where you could transfer but not check.  (I'm also planning to implement hdfs stageout using `hadoop dfs -put`.)  Then I don't think there will be any reason to pass through cases where we didn't manage to measure the remote file size, so if no method succeeds, that will be treated as a failure (e.g. `check_output` returns `False`).

I know you have both moved on, but @matz-e or @annawoodard, if you have a moment to look over this plan and see any flaws with it, please let me know.

**Comment by matz-e** [comment]
@klannon I think that your plan sounds reasonable. Considering the `check_output` implementation, I think that we also once upon a time had the option to run with WorkQueue transferring all the output files, thus having a positive `check_output` when no valid output transfer happens would still be good to have.

**Comment by khurtado** [comment]
For the records, it seems `WMCore` and `CRAB` use mostly `exit codes` coming from their different [stageout implementations](https://github.com/dmwm/CRABServer/blob/master/scripts/cmscp.py#L535), but most of these implementations either rely on the transfer tools checksum verifications, like with [xrdcp](https://github.com/dmwm/WMCore/blob/c2f3404c8b66357276b381c7472dacc130ef0887/src/python/WMCore/Storage/Backends/XRDCPImpl.py#L125) or [gfal](https://github.com/dmwm/WMCore/blob/f538b845facfcde26819e453495c59b6103c3ca3/src/python/WMCore/Storage/Plugins/GFAL2Impl.py#L55), or just use the file size for simple copies with [cp](https://github.com/dmwm/WMCore/blob/c2f3404c8b66357276b381c7472dacc130ef0887/src/python/WMCore/Storage/Plugins/CPImpl.py#L39-L42)

With that being said, it seems CRAB actually disabled checksums from [gfal](https://github.com/dmwm/CRABServer/commit/ab36bc2921b4bc019cf26966e9932c68cf3b2e86) because [not all sites can handle it](https://github.com/dmwm/CRABServer/issues/5319)

TLDR;: 
CRAB/WMCore is:
- Trusting `gfal-copy` exit codes with no checksum verification. 
- Trusting `xrdcp` exit codes but using (I think) adler32 algorithm verification for xrdcp (via `xrdcp -C adler32`)
- Using file size comparison for simple `cp` calls
